src/query/executor.rs

```diff
@@ -63,54 +71,104 @@ where
+            do_query(config, gateway, input, |prss, gateway, input| {
+                Box::pin(execute_test_multiply::<crate::ff::Fp31>(
+                    prss, gateway, input,
+                ))
+            })
+        }
+        #[cfg(any(test, feature = "cli", feature = "test-fixture"))]
+        (QueryType::TestMultiply, FieldType::Fp32BitPrime) => {
+            do_query(config, gateway, input, |prss, gateway, input| {
+                Box::pin(execute_test_multiply::<Fp32BitPrime>(prss, gateway, input))
+            })
+        }
+        #[cfg(any(test, feature = "weak-field"))]
+        (QueryType::SemiHonestIpa(ipa_config), FieldType::Fp31) => {
+            do_query(config, gateway, input, move |prss, gateway, input| {
+                let ctx = SemiHonestContext::new(prss, gateway);
+                Box::pin(
+                    IpaQuery::<crate::ff::Fp31, _, _>::new(ipa_config)
+                        .execute(ctx, input)
+                        .then(|res| ready(res.map(|out| Box::new(out) as Box<dyn Result>))),
+                )
+            })
+        }
+        (QueryType::SemiHonestIpa(ipa_config), FieldType::Fp32BitPrime) => {
+            do_query(config, gateway, input, move |prss, gateway, input| {
+                let ctx = SemiHonestContext::new(prss, gateway);
+                Box::pin(
+                    IpaQuery::<Fp32BitPrime, _, _>::new(ipa_config)
+                        .execute(ctx, input)
+                        .then(|res| ready(res.map(|out| Box::new(out) as Box<dyn Result>))),
+                )
+            })
+        }
+        #[cfg(any(test, feature = "weak-field"))]
+        (QueryType::MaliciousIpa(ipa_config), FieldType::Fp31) => {
+            do_query(config, gateway, input, move |prss, gateway, input| {
+                let ctx = MaliciousContext::new(prss, gateway);
+                Box::pin(
+                    IpaQuery::<crate::ff::Fp31, _, _>::new(ipa_config)
+                        .execute(ctx, input)
+                        .then(|res| ready(res.map(|out| Box::new(out) as Box<dyn Result>))),
+                )
+            })
+        }
+        (QueryType::MaliciousIpa(ipa_config), FieldType::Fp32BitPrime) => {
+            do_query(config, gateway, input, move |prss, gateway, input| {
+                let ctx = MaliciousContext::new(prss, gateway);
+                Box::pin(
+                    IpaQuery::<Fp32BitPrime, _, _>::new(ipa_config)
+                        .execute(ctx, input)
+                        .then(|res| ready(res.map(|out| Box::new(out) as Box<dyn Result>))),
+                )
+            })
+        }
+    }
+}
+
+pub fn do_query<F>(
+    config: QueryConfig,
+    gateway: Gateway,
+    input: ByteArrStream,
+    query_impl: F,
+) -> JoinHandle<QueryResult>
+where
+    F: for<'a> FnOnce(
+            &'a PrssEndpoint,
+            &'a Gateway,
+            ByteArrStream,
+        ) -> Pin<Box<dyn Future<Output = QueryResult> + Send + 'a>>
```

Comment:
This box is needed only because having two type parameters for the function and the returned future prevents describing the HRTB for the `'a` lifetime.

Reply:
I'm still amazed at how you can operate with lifetime traits.  I like rust, but hate having to deal with lifetimes.  It's as though the language was designed to suck you in and then trap you in a Kafkaesque nightmare filled with unbalanced quotes.